// @sanity/components/lib/dialogs/stories/fullscreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FullscreenStory = FullscreenStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _portal = require("part:@sanity/components/portal");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var chance = new _chance.default();
var dialogTestContent = {
  minimal: 'minimal',
  paragraph: 'paragraph',
  longtext: 'longtext',
  example: 'example with dialogcontent'
};
var paragraph = chance.paragraph();
var paragraphs = (0, _range2.default)(0, 20).map(i => /*#__PURE__*/_react.default.createElement("p", {
  key: i
}, chance.paragraph()));

function renderFullscreenContent(type) {
  switch (type) {
    case 'paragraph':
      return /*#__PURE__*/_react.default.createElement("p", null, paragraph);

    case 'longtext':
      return /*#__PURE__*/_react.default.createElement("div", null, paragraphs);

    case 'example':
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("h1", null, "With dialog content"), /*#__PURE__*/_react.default.createElement("p", null, paragraph));

    default:
      return 'Minimal';
  }
}

function FullscreenStory() {
  var actions = [{
    index: '1',
    title: 'Default'
  }, {
    index: '4',
    title: 'Secondary',
    kind: 'simple',
    secondary: true
  }];
  var dialogActions = (0, _knobs.boolean)('Include actions', false, 'test') ? actions : [];
  var contentTest = (0, _knobs.select)('content', dialogTestContent, 'minimal');
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/fullscreen",
    propTables: [_fullscreen.default]
  }, /*#__PURE__*/_react.default.createElement(DialogExample, {
    title: (0, _knobs.text)('title', undefined, 'props'),
    onClose: (0, _knobs.boolean)('Has onClose', false, 'test') && (0, _actions.action)('onClose'),
    centered: (0, _knobs.boolean)('centered', false, 'props'),
    color: (0, _knobs.select)('Color', ['default', 'danger', 'success', 'info', 'warning'], undefined, 'props'),
    actions: dialogActions,
    onAction: (0, _actions.action)('onAction')
  }, contentTest && renderFullscreenContent(contentTest)));
}

function DialogExample(props) {
  var children = props.children,
      restProps = _objectWithoutProperties(props, ["children"]);

  var portalRef = (0, _react.useRef)(document.createElement('div'));
  (0, _react.useEffect)(() => {
    portalRef.current.setAttribute('data-portal', '');
    document.body.appendChild(portalRef.current);
    return () => {
      document.body.removeChild(portalRef.current);
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_portal.PortalProvider, {
    element: portalRef.current
  }, /*#__PURE__*/_react.default.createElement(_fullscreen.default, restProps, children));
}
