// @sanity/components/lib/utilities/Poppable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactPopper = require("react-popper");

var _CaptureOutsideClicks = _interopRequireDefault(require("./CaptureOutsideClicks"));

var _Escapable = _interopRequireDefault(require("./Escapable"));

var _Portal = require("./Portal");

var _Stacked = _interopRequireDefault(require("./Stacked"));

var _Poppable = _interopRequireDefault(require("./Poppable.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DEFAULT_MODIFIERS = [{
  name: 'preventOverflow',
  options: {
    boundaryElement: 'viewport'
  }
}]; // @todo: refactor to functional component

class Poppable extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "popperNode", null);

    _defineProperty(this, "setPopperNode", node => {
      this.popperNode = node;
    });

    _defineProperty(this, "handleClickOutside", ev => {
      if (!this.popperNode || !ev.target) {
        return;
      }

      if (!this.popperNode.contains(ev.target)) {
        var onClickOutside = this.props.onClickOutside;
        if (onClickOutside) onClickOutside(ev);
      }
    });
  }

  render() {
    var _this$props = this.props,
        onEscape = _this$props.onEscape,
        onClickOutside = _this$props.onClickOutside,
        children = _this$props.children,
        referenceClassName = _this$props.referenceClassName,
        _this$props$modifiers = _this$props.modifiers,
        modifiers = _this$props$modifiers === void 0 ? DEFAULT_MODIFIERS : _this$props$modifiers,
        _this$props$placement = _this$props.placement,
        placement = _this$props$placement === void 0 ? 'bottom-start' : _this$props$placement,
        popperClassName = _this$props.popperClassName,
        referenceElement = _this$props.referenceElement; // Undefined referenceElement causes Popper to think it is defined

    var popperPropHack = {};

    if (referenceElement) {
      popperPropHack.referenceElement = referenceElement;
    }

    return /*#__PURE__*/_react.default.createElement(_reactPopper.Manager, null, !referenceElement && /*#__PURE__*/_react.default.createElement(_reactPopper.Reference, null, (_ref) => {
      var ref = _ref.ref;
      return /*#__PURE__*/_react.default.createElement("div", {
        ref: ref,
        className: referenceClassName
      });
    }), children && /*#__PURE__*/_react.default.createElement(_Portal.Portal, null, /*#__PURE__*/_react.default.createElement(_Stacked.default, null, isActive => /*#__PURE__*/_react.default.createElement("div", {
      className: _Poppable.default.portal
    }, /*#__PURE__*/_react.default.createElement(_reactPopper.Popper, _extends({
      innerRef: this.setPopperNode,
      modifiers: modifiers,
      placement: placement
    }, popperPropHack), (_ref2) => {
      var ref = _ref2.ref,
          placementState = _ref2.placement,
          style = _ref2.style;
      return /*#__PURE__*/_react.default.createElement("div", {
        ref: ref,
        style: style,
        "data-placement": placementState,
        className: popperClassName
      }, /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_Escapable.default, {
        onEscape: isActive ? onEscape : undefined
      }), onClickOutside ? /*#__PURE__*/_react.default.createElement(_CaptureOutsideClicks.default, {
        onClickOutside: isActive ? this.handleClickOutside : undefined
      }, children) : children));
    })))));
  }

}

exports.default = Poppable;
